File: gitdot-cli/src/client/methods/question.rs
use anyhow::Result;
use uuid::Uuid;

use gitdot_api::endpoint::question::{
    create_answer::{CreateAnswerRequest, CreateAnswerResponse},
    create_answer_comment::{CreateAnswerCommentRequest, CreateAnswerCommentResponse},
    create_question::{CreateQuestionRequest, CreateQuestionResponse},
    create_question_comment::{CreateQuestionCommentRequest, CreateQuestionCommentResponse},
    get_question::{GetQuestionRequest, GetQuestionResponse},
    get_questions::{GetQuestionsRequest, GetQuestionsResponse},
    update_answer::{UpdateAnswerRequest, UpdateAnswerResponse},
    update_comment::{UpdateCommentRequest, UpdateCommentResponse},
    update_question::{UpdateQuestionRequest, UpdateQuestionResponse},
    vote_answer::{VoteAnswerRequest, VoteAnswerResponse},
    vote_comment::{VoteCommentRequest, VoteCommentResponse},
    vote_question::{VoteQuestionRequest, VoteQuestionResponse},
};

use crate::client::GitdotClient;

impl GitdotClient {
    pub async fn create_question(
        &self,
        owner: &str,
        repo: &str,
        request: CreateQuestionRequest,
    ) -> Result<CreateQuestionResponse> {
        self.post(format!("repository/{}/{}/question", owner, repo), request)
            .await
    }

    pub async fn get_question(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        request: GetQuestionRequest,
    ) -> Result<GetQuestionResponse> {
        self.get(
            format!("repository/{}/{}/question/{}", owner, repo, number),
            request,
        )
        .await
    }

    pub async fn get_questions(
        &self,
        owner: &str,
        repo: &str,
        request: GetQuestionsRequest,
    ) -> Result<GetQuestionsResponse> {
        self.get(format!("repository/{}/{}/questions", owner, repo), request)
            .await
    }

    pub async fn update_question(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        request: UpdateQuestionRequest,
    ) -> Result<UpdateQuestionResponse> {
        self.patch(
            format!("repository/{}/{}/question/{}", owner, repo, number),
            request,
        )
        .await
    }

    pub async fn create_answer(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        request: CreateAnswerRequest,
    ) -> Result<CreateAnswerResponse> {
        self.post(
            format!("repository/{}/{}/question/{}/answer", owner, repo, number),
            request,
        )
        .await
    }

    pub async fn update_answer(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        answer_id: Uuid,
        request: UpdateAnswerRequest,
    ) -> Result<UpdateAnswerResponse> {
        self.patch(
            format!(
                "repository/{}/{}/question/{}/answer/{}",
                owner, repo, number, answer_id
            ),
            request,
        )
        .await
    }

    pub async fn create_answer_comment(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        answer_id: Uuid,
        request: CreateAnswerCommentRequest,
    ) -> Result<CreateAnswerCommentResponse> {
        self.post(
            format!(
                "repository/{}/{}/question/{}/answer/{}/comment",
                owner, repo, number, answer_id
            ),
            request,
        )
        .await
    }

    pub async fn create_question_comment(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        request: CreateQuestionCommentRequest,
    ) -> Result<CreateQuestionCommentResponse> {
        self.post(
            format!("repository/{}/{}/question/{}/comment", owner, repo, number),
            request,
        )
        .await
    }

    pub async fn update_comment(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        comment_id: Uuid,
        request: UpdateCommentRequest,
    ) -> Result<UpdateCommentResponse> {
        self.patch(
            format!(
                "repository/{}/{}/question/{}/comment/{}",
                owner, repo, number, comment_id
            ),
            request,
        )
        .await
    }

    pub async fn vote_question(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        request: VoteQuestionRequest,
    ) -> Result<VoteQuestionResponse> {
        self.post(
            format!("repository/{}/{}/question/{}/vote", owner, repo, number),
            request,
        )
        .await
    }

    pub async fn vote_answer(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        answer_id: Uuid,
        request: VoteAnswerRequest,
    ) -> Result<VoteAnswerResponse> {
        self.post(
            format!(
                "repository/{}/{}/question/{}/answer/{}/vote",
                owner, repo, number, answer_id
            ),
            request,
        )
        .await
    }

    pub async fn vote_comment(
        &self,
        owner: &str,
        repo: &str,
        number: i32,
        comment_id: Uuid,
        request: VoteCommentRequest,
    ) -> Result<VoteCommentResponse> {
        self.post(
            format!(
                "repository/{}/{}/question/{}/comment/{}/vote",
                owner, repo, number, comment_id
            ),
            request,
        )
        .await
    }
}

File: gitdot-web/app/lib/dal/question.ts
import "server-only";

import {
  AnswerResource,
  CommentResource,
  QuestionResource,
  VoteResource,
} from "gitdot-api-ts";
import { z } from "zod";
import {
  authFetch,
  authPatch,
  authPost,
  GITDOT_SERVER_URL,
  handleResponse,
} from "./util";

type CreateQuestionRequest = { title: string; body: string };
type UpdateQuestionRequest = { title: string; body: string };
type CreateAnswerRequest = { body: string };
type UpdateAnswerRequest = { body: string };
type CreateCommentRequest = { body: string };
type UpdateCommentRequest = { body: string };
type VoteRequest = { value: number };

export async function createQuestion(
  owner: string,
  repo: string,
  request: CreateQuestionRequest,
): Promise<QuestionResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question`,
    request,
  );

  return await handleResponse(response, QuestionResource);
}

export async function getQuestion(
  owner: string,
  repo: string,
  number: number,
): Promise<QuestionResource | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}`,
  );

  return await handleResponse(response, QuestionResource);
}

export async function updateQuestion(
  owner: string,
  repo: string,
  number: number,
  request: UpdateQuestionRequest,
): Promise<QuestionResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}`,
    request,
  );

  return await handleResponse(response, QuestionResource);
}

export async function getQuestions(
  owner: string,
  repo: string,
): Promise<QuestionResource[] | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/questions`,
  );

  return await handleResponse(response, z.array(QuestionResource));
}

export async function createAnswer(
  owner: string,
  repo: string,
  number: number,
  request: CreateAnswerRequest,
): Promise<AnswerResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/answer`,
    request,
  );

  return await handleResponse(response, AnswerResource);
}

export async function updateAnswer(
  owner: string,
  repo: string,
  number: number,
  answerId: string,
  request: UpdateAnswerRequest,
): Promise<AnswerResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/answer/${answerId}`,
    request,
  );

  return await handleResponse(response, AnswerResource);
}

export async function createQuestionComment(
  owner: string,
  repo: string,
  number: number,
  request: CreateCommentRequest,
): Promise<CommentResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/comment`,
    request,
  );

  return await handleResponse(response, CommentResource);
}

export async function createAnswerComment(
  owner: string,
  repo: string,
  number: number,
  answerId: string,
  request: CreateCommentRequest,
): Promise<CommentResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/answer/${answerId}/comment`,
    request,
  );

  return await handleResponse(response, CommentResource);
}

export async function updateComment(
  owner: string,
  repo: string,
  number: number,
  commentId: string,
  request: UpdateCommentRequest,
): Promise<CommentResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/comment/${commentId}`,
    request,
  );

  return await handleResponse(response, CommentResource);
}

export async function voteQuestion(
  owner: string,
  repo: string,
  number: number,
  request: VoteRequest,
): Promise<VoteResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/vote`,
    request,
  );

  return await handleResponse(response, VoteResource);
}

export async function voteAnswer(
  owner: string,
  repo: string,
  number: number,
  answerId: string,
  request: VoteRequest,
): Promise<VoteResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/answer/${answerId}/vote`,
    request,
  );

  return await handleResponse(response, VoteResource);
}

export async function voteComment(
  owner: string,
  repo: string,
  number: number,
  commentId: string,
  request: VoteRequest,
): Promise<VoteResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${owner}/${repo}/question/${number}/comment/${commentId}/vote`,
    request,
  );

  return await handleResponse(response, VoteResource);
}
